enhancements/update/standalone-worker-surge.md

```diff
@@ -0,0 +1,295 @@
+managing all aspects of a cluster update. They are operating a standalone OpenShift cluster.
+
+1. The cluster lifecycle administrator desires to ensure that there is sufficient worker node capacity during
+   updates to handle graceful termination of pods and rescheduling of workloads.
+2. They want to avoid other causes of drain stalls by limiting the amount of time permitted for any drain operation.
+3. They configure worker `MachineConfigPools` on the cluster with a `MaxSurge`.
+   value that will bring additional worker node capacity online for the duration of an update.
+4. They configure worker `MachineConfigPools` on the cluster with a `NodeDrainTimeout` value of 30 minutes to 
+   limit the amount of time non-capacity related draining issues can stall the overall update.
+
+### API Extensions
+
+#### API Overview
+The Standalone `MachineConfigPool` custom resource is updated to include new update strategies (one of which 
+supports`MaxSurge`) and `NodeDrainTimeout` semantics identical to HCP's `NodePool`.
+
+Documentation for these configuration options can be found in HyperShift's API reference: 
+- https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.NodePoolManagement exposes `MaxSurge`.
+- https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.NodePoolSpec exposes `NodePoolTimeout`
+
+Example `MachineConfigPool` including both `NodeDrainTimeout` and a `MaxSurge` setting:
+```yaml
+kind: MachineConfigPool
+spec:
+  # Existing spec fields are not shown.    
+
+  # Adopted from NodePool to create consistency and further our goal
+  # to improve the reliability of worker updates.
+  nodeDrainTimeout: 10m
+      
+  # New policy analog to NodePool.NodePoolManagement.
+  machineManagement:
+    upgradeType: "Replace"
+    replace:
+      strategy: "RollingUpdate"
+      rollingUpdate:
+        maxUnavailable: 0
+        maxSurge: 4
+```
+
+Like HCP, `UpgradeType` will support:
+- `InPlace` where no additional nodes are brought online to support draining workloads.
+- `Replace` where new nodes will be brought online with `MaxSurge` support.
+
+#### InPlace Update Type
+
+The `InPlace` update type is similar to `MachineConfigPools` traditional behavior where a 
+user can configure the `MaxUnavailable` nodes. This approach assumes the number (or percentage)
+of nodes specified by `MaxUnavailable` can be drained simultaneously with workloads finding
+sufficient resources on other nodes to avoid stalled drains.
+
+```yaml
+kind: MachineConfigPool
+spec:
+  # Existing spec fields are not shown.    
+
+  machineManagement:
+    upgradeType: "InPlace"
+    inPlace:
+      maxUnavailable: 10%
+```
+
+Nodes are rebooted after they are drained.
+
+#### Replace Update Type
+
+The `Replace` update type removes old machines and replaces them with new instances. It supports
+two strategies:
+- `OnDelete` where a new machine is brought online only after the old machine is deleted.

```

Comment:
You may want to clarify here that "deleted" means marked for deletion, doesn't mean that the node is gone away. Eg the cluster might not have space to move the evicted pods so a new node is needed to come up before this deleted one can go away

Reply:
According to https://github.com/kubernetes-sigs/cluster-api/pull/4346 , OnDelete in CAPI implies:
> The machineDeployment controller waits for the machine to get deleted completely and then will proceed to provision the new replica with the new configuration.

This can absolutely lead to wedging in a cluster with insufficient capacity. The main use case I can see here is if you are truly limited to N nodes by your provider and attempts at N+1 would fail indefinitely. `OnDelete` with a drain timeout achieves an update (with disruption) in this scenario, but you can at least get through the update.

Reply:
Ack, that is not how I remember that working... Hmm, wonder if that changed 🤔 

Then yes, fair with the drain timeout.

Out of interest, what would you expect as an acceptable range of values for the Drain timeout? What if I tried to set it to, 1 year?

Reply:
> Out of interest, what would you expect as an acceptable range of values for the Drain timeout? What if I tried to set it to, 1 year?

Between 1 second (admin just wants the update to unfold immediately) and int64 seconds (grace period  data type podspec). If podspec doesn't have a perspective on what is reasonable, I don't think we should.